### DISCUSSION TITLE
Imbalanced Data in KNN

### DISCUSSION THREAD

**Discussion by eeinnew** [teamalgoritma/community#123] [Q&A] [answered]
Hello, I have some trouble in handling Imbalanced Data in KNN, 
so the ratio between 2 levels is 88% and 12% (which I decided to use upsampling) to prevent information losses. 
However, I am quite confused about the code

```
dmy_train <- dmy[index,2:14]
dmy_test <- dmy[-index,2:14]

dmy_train_label <- dmy[index,1]
dmy_test_label <- dmy[-index,1]
```
this code below contain error as there is no more 'y.yes' column (has been subsetted in the chunk above)
```
dmy_train <- upSample(x = dmy_train %>% 
                           select(-y.yes),
                          y = dmy_train$y.yes,
                         yname = "y.yes")
```

![image](https://user-images.githubusercontent.com/81936292/171538482-f63c04ec-aaf6-4325-8bb8-495f088f6b41.png)

**Comment by ysitta** [comment]
Hi @eeinnew,

The error happens because you already separate the predictor (dmy_train) and label (dmy_label). My suggestion is, you can do the upsample first for all the data, then separate by predictor and label.

Here's the code:

`dmy_train <- dmy[index,]`
`dmy_test <- dmy[-index,]
`
do the upsampling

`dmy_train <- upSample(x = dmy_train %>% 
                           select(-y.yes),
                          y = dmy_train$y.yes,
                         yname = "y.yes")`

then, separate the predictor and label

`dmy_train_pred <- dmy_train[,2:14]`
`dmy_train_label <- dmy_train[,1]`

If you find this answer is helpful, please mark as answered ya :)

**Comment by eeinnew** [comment]
```
dmy_train_pred <- dmy_train[,1:47]
dmy_test_pred<-dmy_test[,1:47]
dmy_train_label <- dmy_train[,48]
dmy_test_label<-dmy_test[,48]
```
```

library(class) 

dmy_pred <- knn(train = dmy_train_pred,
                   test = dmy_test_pred,
                   cl = dmy_train_label,
                   k = 206)

```

output : 
Error in knn(train = dmy_train_pred, test = dmy_test_pred, cl = dmy_train_label,  : 
  too many ties in knn

**Comment by NabiilahArdini** [comment, marked as answer]
Halo @eeinnew , 

Dari error message terakhir (too many ties in knn), sepertinya sudah berupa pertanyaan lanjutan dari pertanyaan yang pertama. Namun permasalah itu sepertinya berhubungan dengan data yang mengandung terlalu banyak prediktor yang discrete/nilai bulat yang rendah, sehingga algoritma kesulitan memilih data sebagai tetangga terdekat. 

Berikut referensi diskusinya: https://stats.stackexchange.com/questions/32905/rknn-error-too-many-ties-in-knn

Mungkin Anda bisa mencoba algoritma lain bila problem ini masih beralan dan data sudah memang tepat seperti itu untuk tipe datanya.
